src/app/catalog/product/components/product-master/product-master.component.ts

```diff
@@ -3,29 +3,80 @@
  * SPDX-License-Identifier: MIT
  */
 
-import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
+import {
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  OnInit,
+  QueryList,
+  ViewChildren,
+} from '@angular/core';
 import { Store } from '@ngrx/store';
 import { loadProducts } from '@app/catalog/product/store/product.actions';
 import { selectProducts } from '@app/catalog/product/store/product.selectors';
 import { StateWithCatalog } from '@app/catalog/store/catalog.reducer';
 import { navigate } from '@app/shared/navigation/navigation.actions';
+import { ViewportService } from '@app/catalog/recommendation/services/viewport.service';
+import { ProductComponent } from '@app/shared/components/product/product.component';
+import { Product } from '@models/product';
+import { debounceTime } from 'rxjs';
 
 @Component({
   selector: 'app-product-master',
   changeDetection: ChangeDetectionStrategy.OnPush,
   templateUrl: './product-master.component.html',
   styleUrls: ['./product-master.component.scss'],
 })
-export class ProductMasterComponent implements OnInit {
-  products$ = this.store.select(selectProducts);
+export class ProductMasterComponent implements OnInit, AfterViewInit {
+  public products: Product[];
+  private displayableContentSections: boolean[];
 
-  constructor(private store: Store<StateWithCatalog>) {}
+  @ViewChildren(ProductComponent) productChildren: QueryList<ProductComponent>;
 
-  ngOnInit() {
+  private productReceiveHandler = (products: Product[]): void => {
+    // fill the sections - if product.length is 8 -> there are 3 sections. One section contains 3 product
+    this.displayableContentSections = new Array(Math.round(products.length / 3)).fill(false);
+    this.products = products;
+    this.cdr.markForCheck();
+  };
+
+  constructor(
+    private store: Store<StateWithCatalog>,
+    private viewportService: ViewportService,
+    private cdr: ChangeDetectorRef,
+  ) {
+    this.store.select(selectProducts).subscribe(this.productReceiveHandler);
+  }
+
+  public ngOnInit(): void {
     this.store.dispatch(loadProducts());
   }
 
-  loadProductDetails(id: number) {
+  public loadProductDetails(id: number) {
     this.store.dispatch(navigate({ url: `/product/${id}` }));
   }
+
+  public isContentInTheViewport(index: number): boolean {
+    return this.displayableContentSections[index / 3];
+  }
+
+  public ngAfterViewInit(): void {
```

Comment:
why cant we use the on-viewport here?  https://angular.dev/guide/defer#on-viewport

Reply:
because in that case it would not have been possible to monitor and render Paragrapf (p tag) of the element of the last loaded product when it is in the viewport. You should put a template reference on the block under defer and look at the last one, but this is not possible. The viewchildren is not available in the template eather. 
I just wanted to use the other variant of the defer, when a fuction is passed. 
I know it's a bit complicated, but it works :)